### PR TITLE
[SPARK-21544][DEPLOY][test-maven] Tests jar of some module should not upload twice

### PR DESCRIPTION
**Pull request by caneGuy**
## What changes were proposed in this pull request?

**For moudle below:**
common/network-common
streaming
sql/core
sql/catalyst
**tests.jar will install or deploy twice.Like:**
`[DEBUG] Installing org.apache.spark:spark-streaming_2.11/maven-metadata.xml to /home/mi/.m2/repository/org/apache/spark/spark-streaming_2.11/maven-metadata-local.xml
[INFO] Installing /home/mi/Work/Spark/scala2.11/spark/streaming/target/spark-streaming_2.11-2.1.0-mdh2.1.0.1-SNAPSHOT-tests.jar to /home/mi/.m2/repository/org/apache/spark/spark-streaming_2.11/2.1.0-mdh2.1.0.1-SNAPSHOT/spark-streaming_2.11-2.1.0-mdh2.1.0.1-SNAPSHOT-tests.jar
[DEBUG] Skipped re-installing /home/mi/Work/Spark/scala2.11/spark/streaming/target/spark-streaming_2.11-2.1.0-mdh2.1.0.1-SNAPSHOT-tests.jar to /home/mi/.m2/repository/org/apache/spark/spark-streaming_2.11/2.1.0-mdh2.1.0.1-SNAPSHOT/spark-streaming_2.11-2.1.0-mdh2.1.0.1-SNAPSHOT-tests.jar, seems unchanged`
**The reason is below:**
`[DEBUG]   (f) artifact = org.apache.spark:spark-streaming_2.11:jar:2.1.0-mdh2.1.0.1-SNAPSHOT
[DEBUG]   (f) attachedArtifacts = [org.apache.spark:spark-streaming_2.11:test-jar:tests:2.1.0-mdh2.1.0.1-SNAPSHOT, org.apache.spark:spark-streaming_2.11:jar:tests:2.1.0-mdh2.1.0.1-SNAPSHOT, org.apache.spark:spark
-streaming_2.11:java-source:sources:2.1.0-mdh2.1.0.1-SNAPSHOT, org.apache.spark:spark-streaming_2.11:java-source:test-sources:2.1.0-mdh2.1.0.1-SNAPSHOT, org.apache.spark:spark-streaming_2.11:javadoc:javadoc:2.1.0
-mdh2.1.0.1-SNAPSHOT]`

when executing 'mvn deploy' to nexus during release.I will fail since release nexus can not be overrided.

## How was this patch tested?
Execute 'mvn clean install -Pyarn -Phadoop-2.6 -Phadoop-provided -DskipTests'

### CODE REVIEW

File: streaming/pom.xml [hunk 1]
```diff
@@ -126,6 +126,7 @@
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-shade-plugin</artifactId>
         <configuration>
+          <shadedArtifactAttached>true</shadedArtifactAttached>
```
Comment (srowen) [inline streaming/pom.xml]: OK I understand the other changes, but is this related?

**Review by srowen — COMMENTED** [review]
OK I follow now, and I agree. The parent POM defines a test-jar execution for all projects, it seems. Submodules may want to override the phase it binds to. But that needs to match on `<id>` and needs to appear once. I think you are correct.

File: streaming/pom.xml (hunk 1 above)
Comment (caneGuy) [inline streaming/pom.xml]: Actually, this will infect 'attachedArtifacts' in the log i supply。
I will remove this in case misleading.

**Comment by caneGuy** [comment]
Tests fot pyspark failed 4 cases.Error is below:
ImportError: C extension: 'hashtable' not built. If you want to import pandas from the source directory, you may need to run 'python setup.py build_ext --inplace --force' to build the C extensions first.

I am not familiar with pyspark,does this error may be caused by environment?

**Comment by caneGuy** [comment]
cc @srowen Test done,any more problem?Thanks too much.

**Comment by srowen** [comment]
@vanzin do you have any thoughts on this one?

**Comment by caneGuy** [comment]
cc 
@vanzin @srowen thanks!

**Comment by vanzin** [comment]
Can you add `[test-maven]` to the PR title so it builds with maven?

**Comment by srowen** [comment]
Ping @caneGuy -- adding `[test-maven]` will let us also verify this passes the Maven build

**Comment by caneGuy** [comment]
Sorry for reply so late since these days i have some personal things to handle @srowen @vanzin.Done for comment now!Is it right?

**Comment by srowen** [comment]
Merged to master

**Comment by caneGuy** [comment]
Thanks @srowen @vanzin .

**Comment by linhong-intel** [comment]
@caneGuy Thanks for you fixing. I'm facing this problem as well and your fix works very well.
But the streaming sub-module still generate test jar twice same as your PR description. How did you fix that problem?
I'm working on spark 2.3.2 release

**Comment by mattf-apache** [comment]
@caneGuy , @LinhongLiu , I'm having exactly the problem described in SPARK-21544 with regard to spark-streaming , in Spark 2.4.0.  Did either of you find a solution for it?  Obviously the changes in PR #18745 fixed the problem with sql/catalyst and sql/core, but did not change streaming.  Any help would be greatly appreciated.

**Comment by LinhongLiu** [comment]
@mattf you can remove the shaded test jar in streaming/pom.xml
I checked the shaded test jar is same to the original one.

**Comment by mattf-apache** [comment]
The remaining problem with spark-streaming is due to MSHADE-259, which I'm in process of providing a PR for.  I was unable to find a work-around that could be done from within spark project poms.  
For those users needing to deploy shaded test jars to internal repos, hopefully the maven-shade-plugin team will publish a fixed release soon.  Until then, you might consider using a private build of maven-shade-plugin, with the patch from MSHADE-259.  It turns out to be pretty simple to do.

**Comment by mattf-apache** [comment]
Thanks @LinhongLiu
